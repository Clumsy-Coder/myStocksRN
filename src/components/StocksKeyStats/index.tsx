import React from 'react';
import { Text, H2, H3, View } from 'native-base';
import { StyleSheet } from 'react-native';

import { DataDomain } from '@redux/Stocks/Types';

type OwnProps = {
  data: DataDomain.Quote;
};

const styles = StyleSheet.create({
  root: {
    flex: 1,
  },
  statsContainer: {
    flexDirection: 'row',
    marginHorizontal: 20,
    marginTop: 10,
    flexWrap: 'wrap',
    alignItems: 'flex-start',
  },
  item: {
    paddingBottom: 10,
    flexBasis: '50%', // make flexbox fit two key stats per row.
  },
  boldFont: {
    fontWeight: 'bold',
  },
});

// key stats to display. Will display data in two columns.
// every index % 2 === 0 is a new row.
const keyStats: { title: string; key: keyof DataDomain.Quote }[] = [
  {
    title: 'Previous close',
    key: 'previousClose',
  },
  {
    title: 'Previous volume',
    key: 'previousVolume',
  },
  {
    title: 'Open',
    key: 'open',
  },
  {
    title: 'Close',
    key: 'close',
  },
  {
    title: 'High',
    key: 'high',
  },
  {
    title: 'Low',
    key: 'low',
  },
  {
    title: '52 Week High',
    key: 'week52High',
  },
  {
    title: '52 Week Low',
    key: 'week52Low',
  },
  {
    title: 'Market capital',
    key: 'marketCap',
  },
  {
    title: 'PE ratio',
    key: 'peRatio',
  },
  {
    title: 'Avg total volume',
    key: 'avgTotalVolume',
  },
  {
    title: 'Volume',
    key: 'volume',
  },
];

const StocksKeyStats: React.FC<OwnProps> = ({ data }: OwnProps) => (
  <View style={styles.root}>
    <H2 style={styles.boldFont}>Key stats</H2>
    <View style={styles.statsContainer}>
      {keyStats.map((cur) => (
        <View key={`keyStat-${cur.title}`} style={styles.item}>
          <H3 style={styles.boldFont}>{cur.title}</H3>
          <Text>
            {data[cur.key] === (undefined || null) ? 'N/A' : Number(data[cur.key]).toLocaleString()}
            {/* NOTE: cannot unit test if data[cur.key] === (undefined || null)
             * This due to the way DataDomain.Quote was designed. It assumes the data is always available.
             * Changing it to allow nullables will require refactor for Reducer, selectors to handle nullables
             */}
          </Text>
        </View>
      ))}
    </View>
  </View>
);

export default StocksKeyStats;
