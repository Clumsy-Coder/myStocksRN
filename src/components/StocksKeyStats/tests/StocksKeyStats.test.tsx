import 'react-native';
import React from 'react';
import { shallow } from 'enzyme';

import StocksKeyStats from 'src/components/StocksKeyStats';
import * as testdata from 'jest.testdata';

describe('<StocksKeyStats />', () => {
  it('Renders correctly', () => {
    const wrapper = shallow(<StocksKeyStats data={testdata.stockQuoteData1} />);

    expect(wrapper).toMatchSnapshot();
  });
});
