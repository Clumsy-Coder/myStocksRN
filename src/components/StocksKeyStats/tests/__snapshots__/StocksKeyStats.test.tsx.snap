// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StocksKeyStats /> Renders correctly 1`] = `
<Styled(ViewNB)
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Styled(H2)
    style={
      Object {
        "fontWeight": "bold",
      }
    }
  >
    Key stats
  </Styled(H2)>
  <Styled(ViewNB)
    style={
      Object {
        "alignItems": "flex-start",
        "flexDirection": "row",
        "flexWrap": "wrap",
        "marginHorizontal": 20,
        "marginTop": 10,
      }
    }
  >
    <Styled(ViewNB)
      key="keyStat-Previous close"
      style={
        Object {
          "flexBasis": "50%",
          "paddingBottom": 10,
        }
      }
    >
      <Styled(H3)
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        Previous close
      </Styled(H3)>
      <Styled(Text)>
        124.16
      </Styled(Text)>
    </Styled(ViewNB)>
    <Styled(ViewNB)
      key="keyStat-Previous volume"
      style={
        Object {
          "flexBasis": "50%",
          "paddingBottom": 10,
        }
      }
    >
      <Styled(H3)
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        Previous volume
      </Styled(H3)>
      <Styled(Text)>
        2,865,362
      </Styled(Text)>
    </Styled(ViewNB)>
    <Styled(ViewNB)
      key="keyStat-Open"
      style={
        Object {
          "flexBasis": "50%",
          "paddingBottom": 10,
        }
      }
    >
      <Styled(H3)
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        Open
      </Styled(H3)>
      <Styled(Text)>
        110
      </Styled(Text)>
    </Styled(ViewNB)>
    <Styled(ViewNB)
      key="keyStat-Close"
      style={
        Object {
          "flexBasis": "50%",
          "paddingBottom": 10,
        }
      }
    >
      <Styled(H3)
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        Close
      </Styled(H3)>
      <Styled(Text)>
        130
      </Styled(Text)>
    </Styled(ViewNB)>
    <Styled(ViewNB)
      key="keyStat-High"
      style={
        Object {
          "flexBasis": "50%",
          "paddingBottom": 10,
        }
      }
    >
      <Styled(H3)
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        High
      </Styled(H3)>
      <Styled(Text)>
        140
      </Styled(Text)>
    </Styled(ViewNB)>
    <Styled(ViewNB)
      key="keyStat-Low"
      style={
        Object {
          "flexBasis": "50%",
          "paddingBottom": 10,
        }
      }
    >
      <Styled(H3)
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        Low
      </Styled(H3)>
      <Styled(Text)>
        100
      </Styled(Text)>
    </Styled(ViewNB)>
    <Styled(ViewNB)
      key="keyStat-52 Week High"
      style={
        Object {
          "flexBasis": "50%",
          "paddingBottom": 10,
        }
      }
    >
      <Styled(H3)
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        52 Week High
      </Styled(H3)>
      <Styled(Text)>
        158.75
      </Styled(Text)>
    </Styled(ViewNB)>
    <Styled(ViewNB)
      key="keyStat-52 Week Low"
      style={
        Object {
          "flexBasis": "50%",
          "paddingBottom": 10,
        }
      }
    >
      <Styled(H3)
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        52 Week Low
      </Styled(H3)>
      <Styled(Text)>
        90.56
      </Styled(Text)>
    </Styled(ViewNB)>
    <Styled(ViewNB)
      key="keyStat-Market capital"
      style={
        Object {
          "flexBasis": "50%",
          "paddingBottom": 10,
        }
      }
    >
      <Styled(H3)
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        Market capital
      </Styled(H3)>
      <Styled(Text)>
        108,740,133,240
      </Styled(Text)>
    </Styled(ViewNB)>
    <Styled(ViewNB)
      key="keyStat-PE ratio"
      style={
        Object {
          "flexBasis": "50%",
          "paddingBottom": 10,
        }
      }
    >
      <Styled(H3)
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        PE ratio
      </Styled(H3)>
      <Styled(Text)>
        12.05
      </Styled(Text)>
    </Styled(ViewNB)>
    <Styled(ViewNB)
      key="keyStat-Avg total volume"
      style={
        Object {
          "flexBasis": "50%",
          "paddingBottom": 10,
        }
      }
    >
      <Styled(H3)
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        Avg total volume
      </Styled(H3)>
      <Styled(Text)>
        4,866,120
      </Styled(Text)>
    </Styled(ViewNB)>
    <Styled(ViewNB)
      key="keyStat-Volume"
      style={
        Object {
          "flexBasis": "50%",
          "paddingBottom": 10,
        }
      }
    >
      <Styled(H3)
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        Volume
      </Styled(H3)>
      <Styled(Text)>
        2,865,320
      </Styled(Text)>
    </Styled(ViewNB)>
  </Styled(ViewNB)>
</Styled(ViewNB)>
`;
